src/babylon/shaders/dvr.fragment.glsl

```diff
@@ -1,26 +1,173 @@
 precision highp float;
 precision highp usampler3D;
+precision highp int;
 
 uniform usampler3D cubeTex;
 uniform float fovy;
+uniform lowp usampler3D voxelCache;
+uniform lowp usampler3D pageTable;
+uniform lowp usampler3D pageDirectory;
 
 in vec3 frontPos;
 in vec3 rayDir;
 
 //layout(location = 0) out vec4 glFragColor;
 layout(location = 1) out uint glFragSegID;
 layout(location = 2) out vec4 glFragDepth;
+layout(location = 3) out uvec4 glCacheState;
 
 const vec3 ZERO3 = vec3(0.0);
 const vec4 ZERO4 = vec4(0.0);
 const vec3 ONE3 = vec3(1.0);
 const vec4 ONE4 = vec4(1.0);
+const vec4 RED = vec4(1.0, 0.0, 0.0, 1.0);
+const vec4 GREEN = vec4(0.0, 1.0, 0.0, 1.0);
+const vec4 BLUE = vec4(0.0, 0.0, 1.0, 1.0);
 
 const float SQRT2INV = 0.70710678118;
 const float SQRT3INV = 0.57735026919;
 
 const float STEPSIZE = 0.001;
 
+// TODO move this to uniforms!
+const bool DEBUG = true;
+const float DATASET_SIZE = 4096.0;
+const float BLOCK_SIZE_VOXEL = 32.0;
+const float BLOCK_SIZE_PAGE = 32.0;
+const float TABLE_SIZE_VOXEL = 16.0;
+const float TABLE_SIZE_PAGE = 16.0;
+const float TOTAL_SIZE_PAGE_DIRECTORY = 4.0;
+const float TOTAL_SIZE_PAGE_TABLES = BLOCK_SIZE_PAGE * TABLE_SIZE_VOXEL;
+const float TOTAL_SIZE_VOXEL_TABLE = BLOCK_SIZE_VOXEL * TABLE_SIZE_PAGE;
+const uint EMPTY = uint(0);
+const uint MAPPED = uint(1);
+const uint NOT_MAPPED = uint(2);
+
+const uint PAGE_DIRECTORY = uint(4);
+const uint PAGE_TABLE = uint(8);
+const uint VOXEL_BLOCK = uint(12);
+
+// --------------------- BEGIN Voxel Lookup ----------------------------
+
+/*
+ * Get the Voxel Coordinates from Normalized Device Coordinates
+ *
+ * @param position Coordinates in NDC: [0,1]
+ *
+ * @return coordinates in voxel space
+ */
+vec3 getVoxelCoordinates(vec3 position) {
+  return position * DATASET_SIZE;
+}
+
+/*
+ * Look up an entry from the Page Directory
+ *
+ * @param positionVoxel The actual raw position in the dataset we are looking at in voxel space
+ * 
+ * @return PageDirectoryEntry (RGBA8 uint) :
+ *                        R - x coordinate in Page Table
+ *                        G - y coordinate in Page Table
+ *                        B - z coordinate in Page Table
+ *                        A - Mapping flag indicates mapping status
+ */
+uvec4 getDirectoryEntry(vec3 positionVoxel) {
+  // Get bits representing page directory ( lop off page table and voxel table bits )
+  vec3 positionPageDirectory = positionVoxel / BLOCK_SIZE_PAGE / BLOCK_SIZE_VOXEL;
+
+  // Use bits to index into page directory
+  return textureLod(pageDirectory, positionPageDirectory / TOTAL_SIZE_PAGE_DIRECTORY, 0.0);
```

Comment:
There is another function `texelFetch(sampler, p, lod)` that takes an ivec of unnormalized coordinates.